op-node/rollup/derive/blob_data_source.go

```diff
@@ -0,0 +1,172 @@
+	data, hashes := dataAndHashesFromTxs(txs, &ds.dsCfg, ds.batcherAddr)
+
+	if len(hashes) == 0 {
+		// there are no blobs to fetch so we can return immediately
+		return data, nil
+	}
+
+	// download the actual blob bodies corresponding to the indexed blob hashes
+	blobs, err := ds.blobsFetcher.GetBlobs(ctx, ds.ref, hashes)
+	if errors.Is(err, ethereum.NotFound) {
+		// If the L1 block was available, then the blobs should be available too. The only
+		// exception is if the blob retention window has expired, which we will ultimately handle
+		// by failing over to a blob archival service.
+		return nil, NewResetError(fmt.Errorf("%w: failed to fetch blobs", err))
+	} else if err != nil {
+		return nil, NewTemporaryError(fmt.Errorf("%w: failed to fetch blobs", err))
+	}
+
+	// go back over the data array and populate the blob pointers
+	if err := fillBlobPointers(data, blobs); err != nil {
+		// this shouldn't happen unless there is a bug in the blobs fetcher
+		return nil, NewResetError(fmt.Errorf("%w: failed to fill blob pointers", err))
+	}
+	return data, nil
+}
+
+// dataAndHashesFromTxs extracts calldata and datahashes from the input transactions and returns them. It
+// creates a placeholder blobOrCalldata element for each returned blob hash that must be populated
+// by fillBlobPointers after blob bodies are retrieved.
+func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address) ([]blobOrCalldata, []eth.IndexedBlobHash) {
+	data := []blobOrCalldata{}
+	var hashes []eth.IndexedBlobHash
+	blobIndex := 0 // index of each blob in the block's blob sidecar
+	for _, tx := range txs {
+		// skip any non-batcher transactions
+		if !isValidBatchTx(tx, config.l1Signer, config.batchInboxAddress, batcherAddr) {
+			blobIndex += len(tx.BlobHashes())
+			continue
+		}
+		// handle non-blob batcher transactions by extracting their calldata
+		if tx.Type() != types.BlobTxType {
+			calldata := eth.Data(tx.Data())
+			data = append(data, blobOrCalldata{nil, &calldata})
+			continue
+		}
+		// handle blob batcher transactions by extracting their blob hashes, ignoring any calldata.
+		if len(tx.Data()) > 0 {
+			log.Warn("blob tx has calldata, which will be ignored", "txhash", tx.Hash())
+		}
+		for _, h := range tx.BlobHashes() {
+			idh := eth.IndexedBlobHash{
+				Index: uint64(blobIndex),
+				Hash:  h,
+			}
+			hashes = append(hashes, idh)
+			data = append(data, blobOrCalldata{nil, nil}) // will fill in blob pointers after we download them below
+			blobIndex += 1
+		}
+	}
+	return data, hashes
+}
+
+// fillBlobPointers goes back through the data array and fills in the pointers to the fetched blob
+// bodies. There should be exactly one placeholder blobOrCalldata element for each blob, otherwise
+// error is returned.
+func fillBlobPointers(data []blobOrCalldata, blobs []*eth.Blob) error {
+	blobIndex := 0
+	for i := range data {
+		if data[i].calldata != nil {
+			continue
+		}
+		if blobIndex >= len(blobs) {
```

Comment:
Thoughts on defining these errors at the top of the file as variables and then wrapping them with contextual info using `fmt.Errorf`? Then we can test for specific errors in this function

Reply:
none of these errors should ever be triggered unless someone introduces a logic bug in other parts of the implementation, so not sure it's worth it?  Happy to do it though if you think it's valuable.

Reply:
Agree with @roberto-bayardo here, these are unexpected errors where no special error handling logic would be implemented at the calling site, so not good candidates for dedicated error types.